[Docs] CogManagerUI user guide (#4152)
Cog-Creators/Red-DiscordBot: docs/cog_guides/cog_manager_ui.rst

.. _cogmanagerui:

==============
Cog Manager UI
==============

This is the cog guide for the core cog. You will
find detailed docs about usage and commands.

``[p]`` is considered as your prefix.

.. note:: This cog is not like the other cogs. It is loaded by default, not
    included in the cogs paths and it cannot be unloaded. It contains needed

>     commands for cogs management.

Reviewer: ```suggestion
    commands for cog management.
```

.. _cogmanaerui-usage:

-----
Usage
-----

This cog allows you to manage your cogs and where you can install them. Unlike
V2, which had a ``cogs`` folder where everything was installed, you can
install V3 cogs everywhere, and also make them cross-compatible with other
instances!

The most basic command is :ref:`paths <cogmanagerui-command-paths>`, which
will list you all of the currently set paths.

You can add a path by using the :ref:`addpath <cogmanagerui-command-addpath>`
command. All cogs in that path will be available for the bot and listed in
the :ref:`cogs <cogmanagerui-command-cogs>`. You can then load

> or unload them.

Reviewer 2: I feel like it might be helpful to clarify that this is **only** for installing cogs from files, and that 3rd party cogs found on cog lists **do not** need to be installed this way, and should be installed with downloader instead.

.. TODO add ref to load and unload commands

.. _cogmanagerui-usage-installation:

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
How to install a local package without using downloader
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Let's suppose you made a cog request on the `cog board <https://cogboard.red>`_
and now you want to add your own cog to Red. You should have a folder that
looks like this:

.. image:: ../.resources/cog_manager_ui/custom-cog-example.png

You will first need to add a cog path to your instance. For that, use the
:ref:`addpath <cogmanagerui-command-addpath>` command with a new directory.

Create a folder somewhere (should stay accessible) and copy its path. A path
looks like this:

*   Windows: ``C:\Users\username\Documents\MyCogs``
*   macOS: ``/Users/username/Documents/MyCogs``
*   Linux: ``/home/username/Documents/MyCogs``

You can now use the command we talked about before: type ``[p]addpath
<your_path>``.

> .. attention:: A path shouldn't have spaces in it. If it does, add quotation
>     marks, or a backslash before the space.

Reviewer 2: ```suggestion
.. attention:: A path shouldn't have spaces in it. If it does, add quotation
    marks around the path, or a backslash before the space.
```

>     marks, or an antislash before the space.

Reviewer: ```suggestion
    marks, or a backslash before the space.
```
You use backslash later on, and I would call \ a backslash, not an antislash.

> In that ``MyCogs`` folder, you can drop your package. You should now have
> something like this:

Reviewer 2: ```suggestion
In that ``MyCogs`` folder, you can drop your cog folder. You should now have
something that looks like this:
```

.. image:: ../.resources/cog_manager_ui/cog-path.png

Now if you type ``[p]cogs``, your new cog should be listed, and you will be
able to load it!

.. _cogmanagerui-commands:

--------
Commands
--------

.. note:: The whole cog is locked to the
    :ref:`bot owner <getting-started-permissions>`. If you are not the owner
    of the instance, you can ignore this.

.. _cogmanagerui-command-cogs:

^^^^
cogs
^^^^

**Syntax**

.. code-block:: none

    [p]cogs

**Description**

> Return a list of loaded and unloaded cogs on the bot.

Reviewer: ```suggestion
Returns a list of loaded and unloaded cogs on the bot.
```

> Cogs are unloaded by default, this is where you can find your cogs if you

Reviewer: ```suggestion
Cogs are unloaded by default. This is where you can find your cogs if you
```
I don't feel like these are (relatively) closely related so probably best to separate them with a full stop.

under the path will be added to the list of available cogs, listed in
:ref:`cogs <cogmanagerui-command-cogs>`.

**Arguments**

*   ``<path>``: A path that should look like this and point to a folder:

    *   Windows: ``C:\Users\username\Documents\MyCogs``
    *   macOS: ``/Users/username/Documents/MyCogs``
    *   Linux: ``/home/username/Documents/MyCogs``

    Try to avoid paths with spaces. If there are spaces, add a backslash before
    the space on Linux. Add quotation marks around the path if needed.

.. _cogmanagerui-command-removepath:

^^^^^^^^^^
removepath
^^^^^^^^^^

**Syntax**

.. code-block:: none

    [p]removepath <path_number>

**Description**

Removes a path from the list of available paths. Its cogs won't be accessible
anymore.

**Arguments**

*   ``<path_number>``: The number of the path to remove. You can get it with
    the :ref:`paths <cogmanagerui>` command.

.. _cogmanagerui-command-reorderpath:

^^^^^^^^^^^
reorderpath
^^^^^^^^^^^

**Syntax**

.. code-block:: none

    [p]reorderpath <from\_> <to>

**Description**

Reorders the paths listed with the :ref:`paths <cogmanagerui-command-paths>`

> comand. The goal of this command is to allow the discovery of different cogs.

Reviewer: ```suggestion
command. The goal of this command is to allow the discovery of different cogs.
```
